bip-0345.mediawiki

```diff
@@ -255,28 +254,45 @@ where
 ** Amount checks are ultimately done with deferred checks, but this check can help short-circuit obviously invalid spends.
 * Queue a deferred check<ref>'''What is a deferred check and why does this proposal require them for correct script evaluation?''' A deferred check is a validation check that is executed only after all input scripts have been validated, and is based on aggregate information collected during each input's EvalScript run.<br /><br />Currently, the validity of each input is (usually) checked concurrently across all inputs in a transaction. Because this proposal allows batching the spend of multiple vault inputs into a single recovery or withdrawal output, we need a mechanism to ensure that all expected values per output can be summed and then checked. This necessitates the introduction of an "aggregating" set of checks which can only be executed after each input's script is evaluated. Note that similar functionality would be required for batch input validation or cross-input signature aggregation.</ref> that ensures the satoshis for this input's <code>nValue</code> minus <code><revault-amount></code> are included within the output <code>nValue</code> found at <code><trigger-vout-idx></code>.
+
+If none of the conditions fail, the minimally-encoded difference between input satoshi value and <code>trigger-amount</code>, here-after known as <code>residual</code> is left on the stack. Typically this residual will be fed directly into <code>OP_REVAULT</code> or constrained in some way to avoid funds being exfiltrated trivially from the smart contract. Note that this doesn't preclude pushing a negative number onto the stack, as <code>trigger-amount</code> can be larger than input value potentially.
+
+=== <code>OP_REVAULT</code> evaluation ===
+
+<source>
+<revault-vout-idx>
+<revault-amount>
+</source>
+
+where
+
+* <code><revault-vout-idx></code> is an up to 4-byte CScriptNum-encoded number indicating the index of an output which, in conjunction with the trigger output, carries forward the value of this input, and has an identical scriptPubKey to the current input.
+** If this value does not decode to a valid CScriptNum, script execution when spending this output MUST fail and terminate immediately.
+** If this value is greater than or equal to the number of outputs, script execution when spending this output MUST fail and terminate immediately.
+
+* <code><revault-amount></code> is an up to 7-byte CScriptNum-encoded number indicating the number of satoshis being revaulted.
+** If this value does not decode to a valid CScriptNum, script execution when spending this output MUST fail and terminate immediately.
+** If this value is not greater than or equal to 0, script execution when spending this output MUST fail and terminate immediately.
+** If this value is non-zero but <code><revault-vout-idx></code> is negative, script execution when spending this output MUST fail and terminate immediately.
```

Comment:
If going forward with this, I'd change the logic to what sanket suggested:

1) if no revault exists, use `-1` for index
2) if revault amount is non-0, index must not be `-1`